### ISSUE TITLE
script resources are broken when used with Ruby shipping with OSX and a non-root user

### ISSUE THREAD

**Issue by vlovich** [chef/chef#3563]
Install chef using gem with the Ruby bundled with 10.10
Then write a script resource that specifies a user:
```
    bash "something" do
        user my_user
        group my_group
        code "echo hello"
    end
```

This does not work because the temporary directory used to write out the script file is in the root's temporary directory.  However it will be my_user who will be executing this script & thus cannot see inside the temporary directory.  For example, the users temporary directory might be /var/folders/s1/t73ss6cd32z3468ct01fhgr80000gn/T/ but root's will be /var/folders/zz/zyxvpxvq6csfxvn_n0000000000000/T (this is a random folder every boot).

The correct action here is that the script provider needs to copy the temporary file to a temporary directory owned by the user as well as changing the permissions or, alternatively, you could pipe the script directly into the invoked interpreter's stdin.

**Comment by lamont-granquist** [comment]
To clarify this is chef-client running as root, but with the execute resource changing the uid, not running chef-client itself as a non-root user.

Piping to stdin seems like a better solution but more work to overhaul the resource.  If there is a way to identify a tmpdir that we can write to that would be easier code most likely but the stdin solution probably has fewer other edge cases and will just work everywhere.

**Comment by vlovich** [comment]
Technically chef-solo running as root but I think it's the same.  AFAIK, you cannot determine the temporary directory of a user with any kind of API as that would violate the security of the implementation.  You could sudo into the user & read the temporary directory that way (if it can be done at that layer).

**Comment by lamont-granquist** [comment]
Yeah we'd probably need to use mixlib-shellout's (relatively new) ability to get a proper login shell and that would hopefully give us a usable tmpdir and then we could both drop the file and execute as the user.  However, we'd have to use the "login shell" feature to drop the file, and if we wanted to support the current non-login shell use case we'd need to change to the user once to drop the file and then switch back to root and then switch back to the user as a non-login shell.  Which seems really fussy.

**Comment by vlovich** [comment]
What about just using /tmp as the temporary directory when a user is set?

**Comment by rmoriz** [comment]
This is a really ugly bug and stole me a couple of hours today.  😡

**Comment by lamont-granquist** [comment]
The execute resource (and the script resources + bash and other resources which inherit from it) need to be able to use the login shell feature of mixlib-shellout.  With `login true` set on any of those resources then when switching from root to a different user in the execute resource on macos, the TEMP directory should be set to the appropriate directory for the target user rather than the old root user TEMP directory.

**Comment by lamont-granquist** [comment]
another dup of #7690

**Comment by lamont-granquist** [comment]
Oh that one is windows this is the mac one.

**Comment by lamont-granquist** [comment]
I think this issue we'll define as "execute resource needs to support mixlib-shellout login shell" option, which should clean up most of the Unix issues.  So that is the "definition of done" for this one.

I'm leaving #7690 open to track additional issues around how we should handle env variables on windows.

**Comment by phiggins** [comment]
Fixed in #9932.
